### ISSUE TITLE
New linter: oe_structure element without specific id

### ISSUE THREAD

**Issue by Yajo** [OCA/pylint-odoo#300]
Odoo website builder will replace a view if it contains something like this and the user edits that section:

```xml
<div class="oe_structure"/>
```

However, if it is in this other format, instead of creating a COW view that completely replaces the previous one, [it will create an extension view that replaces this specific element in the parent view](https://github.com/odoo/odoo/blob/25edcc89273c389f318c434ce9a9e0b531b8e599/addons/web_editor/models/ir_ui_view.py#L41):

```xml
<div id="oe_structure_SOME_HOOK_HERE" class="oe_structure"/>
```

This provides a more resilient upgrade path because your module can "safely"\* update the base view which contains logic, while user customizations are kept separately.

IMHO this is a good candidate for a linter.

---

\* You know that in Odoo nothing is exactly "safe", but it's better. :shrug:

**Comment by moylop260** [comment]
Hi @Yajo 

What tags should raise an error if an `id` is not defined?

Should it be consider qweb `<template` views or only normal ones or both?

**Comment by Yajo** [comment]
> What tags should raise an error if an `id` is not defined?

Usually any HTML block content, such as div, content, etc. However I think you can ignore the tag, because it would make no sense to add a `oe_structure` class to i.e. a span, i, b, etc., wouldn't even make sense.

So we can assume that if any element has `oe_structure`, it should have an id with that format. Otherwise you should either add the id or remove the class.



> Should it be consider qweb `<template` views or only normal ones or both?

Usually only on templates, but again I think it's not important because outside of templates it wouldn't make sense anyway. Same logic as above.

**Comment by moylop260** [comment]
@yajo 

I looked for `grep -r 'class="oe_structure"' . --include=*.xml -A2` in odoo modules and they are a lot of these cases and the following one only declaring `class`:
 - https://github.com/odoo/odoo/blob/bbf8255bbb758c060ec8f58cd0dc4cc10e14a3aa/addons/event/report/event_event_templates.xml#L55

I'm not totally convinced to do this lint seeing the output

Could you check it, please?

**Comment by yajo** [comment]
Yes, they do that and it's horrible. 😄 

If the div is declared like `<div class="oe_structure">`, Odoo will let the user modify it. Then it will create a copy of the whole view, including user's modifications. If you update the base view, the modified copy will not be updated. If the base view had any logic, it won't update.

If the div is declared like `<div id="oe_structure_SOMETHING" class="oe_structure">`, Odoo will also let the user modify it. But then it will create a new view that extends the parent view with an `<xpath expr="//div[@id='oe_structure_SOMETHING']">`. This is much more maintainable because it lets you update the base view while preserving user's modifications.

Since Odoo always uses this bad practice, it's common to imitate them. That's why the linter would help.

**Comment by moylop260** [comment]
@xmo-odoo 
I think this might interest you
